viewing_party/party.py

```diff
@@ -1,23 +1,142 @@
+        else:
+            genre_dict[movie["genre"]] = 1
+
+    most_common_genre = None
+    most_common_count = 0 
+
+    for movie_genre, count in genre_dict.items():
+        if count > most_common_count:
+            most_common_genre = movie_genre
+            most_common_count = count
+
+    return most_common_genre
+
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    user_watched_list = user_data["watched"].copy()
+    for friend in user_data["friends"]:
+        for friend_watched_movie in friend["watched"]:
+            if friend_watched_movie in user_watched_list:
+                user_watched_list.remove(friend_watched_movie)
+    return user_watched_list
+
+def get_friends_unique_watched(user_data):
+    movies_watched_by_friend_only_list = []
+    for friend in user_data["friends"]:
+        for friend_watched_movie in friend["watched"]:
+            if friend_watched_movie in movies_watched_by_friend_only_list:
+                continue
+            if friend_watched_movie not in user_data["watched"]:
+                    movies_watched_by_friend_only_list.append(friend_watched_movie)
+    
+    return movies_watched_by_friend_only_list
 
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    recommendations = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie in recommendations:
+                continue
+            
+            elif movie in user_data["watched"]:
+                continue
+            
+            elif movie["host"] not in user_data["subscriptions"]:
+                continue
+            
+            else: 
+                recommendations.append(movie)
+
+    return recommendations
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    
+    favorite_genre = get_most_watched_genre(user_data)
+    
+    recommended_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["genre"] != favorite_genre: 
+                continue
+            
+            elif movie in recommended_movies:
+                continue
+
+            elif movie in user_data["watched"]:
+                continue
+
+            else: 
+                recommended_movies.append(movie)
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):
+    recommended_list = []
+    new_list = user_data["favorites"].copy()
+
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            if friend_movie in new_list:
+                new_list.remove(friend_movie)

```

Comment:
Since `remove` has linear time complexity, how can we go about writing the logic so that we append a movie to the list instead?

If you get a list of the movies watched by friends (with a helper method like suggested above) and then check that a user's favorite movies are not in that list, then we could append that movie to a new list like `recommendations`. 

Then you wouldn't need a whole copy of `user_data["favorites"]` and would not incur the linear time complexity of removing an element from a list (you'll have the constant time complexity of `append`).

This might look like:
```python
def get_rec_from_favorites():
    recommendations = []
    friends_watched = get_friends_watched_movies()

    for favorite in user_data["favorites"]:
        if not favorite in friends_watched:
           recommendations.append(favorite)
    return recommendations